### ISSUE TITLE
Implement natural joins

### ISSUE THREAD

**Issue by st-pasha** [h2oai/datatable#1190]
One of the most common types of join in SQL is the [**natural join**](https://en.wikipedia.org/wiki/Join_(SQL)#Natural_join). It is very similar to a *left outer join*, with only semantic difference: the natural join occurs over all columns with the same names in the left and right tables (i.e. no need to specify which columns to join on), and also makes sure that these matching columns occur only once in the resulting table. The wiki page also notes that natural joins are frowned upon by the SQL community, because of this automatic column selection. Good thing we are not modeling our API after SQL; `data.table` had natural joins for a very long time, and everyone seems to be happy about them.

To be more specific, the functionality that I'm going to call a *natural join* is somewhere in-between the traditional natural joins and the left outer joins. Here's what I mean:
- In order to "naturally join" frame `X` to a frame `DT`, we shall write `DT[:, :, join(X)]`;
- Such join is possible only if:
    * `X` frame has a key;
    * A column with the same name & type as `X`'s key exists in `DT`;
- Natural join matches each row in `DT` with 1 or 0 rows in `X` according to the key column (since key columns must be unique);
- The resulting frame has the same number and order of rows as `DT`, and the number of columns is `DT.ncols + X.ncols - X.nkeys`;
- The main difference between our natural join and the SQL natural join is that there are constraints on the right table: it must have a unique key ("primary key" in SQL), and only that key is eligible for column matching. In addition, the uniqueness constraint guarantees that the joined frame will have the same number of rows as the source frame, which arguably makes this join more "natural": the source frame merely gets some extra columns appended with the information from another frame.

TL;DR this issue is about implementing the following new syntax:
```
DT[:, :, join(X)]
```
More advanced types of joins will eventually be available too, see #1080.
